crates/sp-domains-fraud-proof/src/host_functions.rs

```diff
@@ -118,6 +116,38 @@ where
         .map(|ext| ext.encode())
     }
 
+    fn derive_domain_set_code_extrinsic(
+        &self,
+        consensus_block_hash: H256,
+        domain_id: DomainId,
+    ) -> Option<Option<Vec<u8>>> {
+        let maybe_upgraded_runtime = maybe_runtime_upgrade::<_, _, DomainBlock>(
+            &self.consensus_client,
+            consensus_block_hash.into(),
+            domain_id,
+        )
+        .ok()
+        .flatten();
+
+        if let Some(upgraded_runtime) = maybe_upgraded_runtime {
+            let runtime_code = self.get_domain_runtime_code(consensus_block_hash, domain_id)?;
+            let domain_runtime_api_light =
+                RuntimeApiLight::new(self.executor.clone(), runtime_code.into());
+
+            SetCodeConstructor::<DomainBlock>::construct_set_code_extrinsic(
+                &domain_runtime_api_light,
+                // We do not care about the domain hash since this is stateless call into
+                // domain runtime,
+                Default::default(),
```

Comment:
I'm not sure why we don't care. We still have to make a query against some block hash that is not pruned even though the call will not read anything from state. Otherwise how do we know which runtime version to even use to handle the call?

Reply:
I'm not sure I follow what you mean by `which runtime version`. This domain runtime was taken from the consensus state at a specific consensus block and then we proceed to call into this runtime to construct that extrinsic. We can ofcourse pass a valid `domain_block_hash` instead of default but it does not offer any change. Hence passed a default value. Maybe I'm missing something you are seeing ?

Reply:
Hm... I see, so `domain_runtime_api_light` is a special client already pre-instantiated with a specific runtime. Why does it have the same API that requires block hash then? Block hash doesn't seem to be applicable to it.

Reply:
So we have two versions, FullApi, LightApi, ignore the naming please or if you have a better name. Both the versions implement same api to make it more composable. FullApi is used by Operator for Domain block processing and LightApi is a stateless one, can inject specific state if required, is used by Fraud proof verifier on Consensus chain. So we had to pass the block hash. But I agree this is awkard and was introduced during the time when we had system domain. I can make a TODO to revisit these to make it more easier to understand

Reply:
Looks like stateful/stateless API to me then. If you want the same abstraction the same then methods should probably not have block hash, rather what is called `FullApi` should take block hash as constructor argument.

Reply:
Yeap agree. Now that I think about we do not need this over abstration anymore. Will remove this mess in a different pr  once some other PRs lands so that I dont give them any conflicts. Thanks!